nat-lab/tests/test_pq.py

```diff
@@ -372,3 +378,108 @@ async def test_dns_with_pq(
 
         # Expect this to work as well after the secret key change
         await query_dns(client_conn, "google.com")
+
+
+@pytest.mark.parametrize(
+    "setup",
+    [
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1,
+                adapter_type_override=TelioAdapterType.NEP_TUN,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.DOCKER_CONE_CLIENT_1,
+                    nlx_1_limits=ConnectionLimits(1, 2),
+                ),
+                is_meshnet=False,
+            ),
+        ),
+    ],
+)
+async def test_pq_vpn_silent_pq_upgrader(
+    setup: SetupParameters,
+) -> None:
+    async with AsyncExitStack() as exit_stack:
+        setup.features.post_quantum_vpn.handshake_retry_interval_s = 1
+
+        env = await exit_stack.enter_async_context(
+            setup_environment(exit_stack, [setup])
+        )
+
+        client_conn, *_ = [conn.connection for conn in env.connections]
+        client, *_ = env.clients
+
+        wg_server = config.WG_SERVER  # use non PQ server
+        try:
+            await client.connect_to_vpn(
+                str(wg_server["ipv4"]),
+                int(wg_server["port"]),
+                str(wg_server["public_key"]),
+                pq=True,
+                timeout=4,
+            )
+            raise Exception("This shouldn't connect succesfully")
+        except TimeoutError:
+            pass
+
+        await client.disconnect_from_vpn(str(wg_server["public_key"]))
+
+        # now connect to a good behaving PQ server
+        await _connect_vpn_pq(client_conn, client)
+        await ping(client_conn, config.PHOTO_ALBUM_IP)
+
+
+@pytest.mark.parametrize(
+    "alpha_setup_params",
+    [
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1,
+                adapter_type_override=TelioAdapterType.NEP_TUN,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.DOCKER_CONE_CLIENT_1,
+                    nlx_1_limits=ConnectionLimits(1, 2),
+                ),
+                is_meshnet=False,
+            ),
+        ),
+    ],
+)
+async def test_pq_vpn_upgrade_from_non_pq(
+    alpha_setup_params: SetupParameters,
+) -> None:
+    async with AsyncExitStack() as exit_stack:
+        alpha_setup_params.features.post_quantum_vpn.handshake_retry_interval_s = 1
+
+        env = await exit_stack.enter_async_context(
+            setup_environment(exit_stack, [alpha_setup_params])
```

Comment:
you need to add `prepare_vpn = True` here, as after [this PR](https://github.com/NordSecurity/libtelio/pull/949) the container preparation has been changed to be done on-deman. It save quite measurable amount of time, as otherwise containers are being prepared 1000 times or so during test-suite.